### PR TITLE
Save end device location on location solved

### PR DESCRIPTION
**Pull request by adriansmares**
#### Summary

References #4267 

#### Changes

- Minor API typo fixing
- Allow cluster authenticated requests to update end device `locations` field
- Update end device locations when a location solved message is issued

#### Testing

Replay the uplink from the earlier PR (https://github.com/TheThingsNetwork/lorawan-stack/pull/4305) - make sure to remove the EUIs as they will cause you headaches.

The result is that end device locations will be updated, and that the location solved event will still fire. Old locations are maintained, and on uplink the solved location is visible.

<details>

```json
{
  "name": "as.up.data.forward",
  "time": "2021-06-25T12:22:07.542656357Z",
  "identifiers": [
    {
      "device_ids": {
        "device_id": "dev1",
        "application_ids": {
          "application_id": "app1"
        },
        "dev_eui": "1111111111111111",
        "join_eui": "0000000000000000"
      }
    }
  ],
  "data": {
    "@type": "type.googleapis.com/ttn.lorawan.v3.ApplicationUp",
    "end_device_ids": {
      "device_id": "dev1",
      "application_ids": {
        "application_id": "app1"
      },
      "dev_eui": "1111111111111111",
      "join_eui": "0000000000000000"
    },
    "correlation_ids": [
      "as:up:01F91HXGXFVZRX72DXQQ5JF37D",
      "rpc:/ttn.lorawan.v3.AppAs/SimulateUplink:01F91HXGWXBHP0VHGCJ7VHMVJM"
    ],
    "received_at": "2021-06-25T12:22:07.535497442Z",
    "uplink_message": {
      "f_port": 1,
      "frm_payload": "/w==",
      "rx_metadata": [
        {
          "gateway_ids": {
            "gateway_id": "test"
          },
          "rssi": 42,
          "channel_rssi": 42,
          "snr": 4.2
        }
      ],
      "settings": {
        "data_rate": {
          "lora": {
            "bandwidth": 125000,
            "spreading_factor": 7
          }
        }
      },
      "received_at": "0001-01-01T00:00:00Z",
      "locations": {
        "lora-cloud-geolocation-v3-rssi": {
          "latitude": 51.501591,
          "longitude": -0.141631,
          "accuracy": 300,
          "source": "SOURCE_LORA_RSSI_GEOLOCATION"
        },
        "user": {
          "latitude": 51.415139048867665,
          "longitude": 5.50689697265625,
          "source": "SOURCE_REGISTRY"
        }
      }
    },
    "simulated": true
  },
  "correlation_ids": [
    "as:up:01F91HXGXFVZRX72DXQQ5JF37D",
    "rpc:/ttn.lorawan.v3.AppAs/SimulateUplink:01F91HXGWXBHP0VHGCJ7VHMVJM"
  ],
  "origin": "A98BCD2222F1",
  "visibility": {
    "rights": [
      "RIGHT_APPLICATION_TRAFFIC_READ"
    ]
  },
  "unique_id": "01F91HXGXPKSFWDGQX7XTTV439"
}
```

</details>

<details>

```bash
adriansmares@A98BCD2222F1:~/go/src/go.thethings.network/lorawan-stack$ ttn-lw-cli device get app1 dev1 --locations
{
  "ids": {
    "device_id": "dev1",
    "application_ids": {
      "application_id": "app1"
    },
    "dev_eui": "1111111111111111",
    "join_eui": "0000000000000000"
  },
  "created_at": "2021-06-25T11:17:58.272Z",
  "updated_at": "2021-06-25T12:14:46.560Z",
  "locations": {
    "lora-cloud-geolocation-v3-rssi": {
      "latitude": 51.501591,
      "longitude": -0.141631,
      "accuracy": 300,
      "source": "SOURCE_LORA_RSSI_GEOLOCATION"
    },
    "user": {
      "latitude": 51.415139048867665,
      "longitude": 5.50689697265625,
      "source": "SOURCE_REGISTRY"
    }
  }
}
```

</details>

##### Regressions

Testing should cover this.

#### Notes for Reviewers

Locations are solved from the worker pool of the solver packages - the hot path is not affected due to the fact that we retrieve/save the locations in-line.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/identityserver/end_device_registry.go [hunk 1]
```diff
@@ -181,7 +182,9 @@ func (is *IdentityServer) setFullEndDevicePictureURL(ctx context.Context, dev *t
 }
 
 func (is *IdentityServer) updateEndDevice(ctx context.Context, req *ttnpb.UpdateEndDeviceRequest) (dev *ttnpb.EndDevice, err error) {
-	if err = rights.RequireApplication(ctx, req.EndDeviceIdentifiers.ApplicationIdentifiers, ttnpb.RIGHT_APPLICATION_DEVICES_WRITE); err != nil {
+	if clusterauth.Authorized(ctx) == nil {
```
Comment (adriansmares) [inline pkg/identityserver/end_device_registry.go:185]: Another solution would be to allow the cluster authentication to have the `RIGHT_APPLICATION_DEVICES_WRITE` right. I think that's too wide for what we are trying to solve here.

**Review by johanstokking — APPROVED** [review]
AS part LGTM

**Comment by johanstokking** [comment]
Hmm, did I introduce a flake?

**Comment by adriansmares** [comment]
> Hmm, did I introduce a flake?

I think I need to add the end device `Get`/`Update` support to the mock IS.

**Comment by adriansmares** [comment]
Adding that caught another bug - the service name for the frame payload deduction mechanism is not valid per our `locations` validation (`_` is not allowed):

https://github.com/TheThingsNetwork/lorawan-stack/blob/4466080950116b2cc50cecdaef57f9ce0da3a9d3/api/end_device.proto#L444

I have renamed the service to `frm-payload` and added tests for the saving behavior.

File: pkg/identityserver/end_device_registry.go (hunk 1 above)
Comment (htdvisser) [reply pkg/identityserver/end_device_registry.go:185]: Indeed, we could consider that for a future refactoring

**Review by htdvisser — APPROVED** [review]

File: pkg/applicationserver/applicationserver.go [hunk 2]
```diff
@@ -916,14 +916,14 @@ func (as *ApplicationServer) handleUplink(ctx context.Context, ids ttnpb.EndDevi
 		if uplink.Locations == nil {
 			uplink.Locations = make(map[string]*ttnpb.Location)
 		}
-		uplink.Locations["frm_payload"] = loc
+		uplink.Locations["frm-payload"] = loc
```
Comment (johanstokking) [inline pkg/applicationserver/applicationserver.go:919]: Thanks, this is something I didn't realize

**Review by johanstokking — APPROVED** [review]

File: pkg/applicationserver/applicationserver.go (hunk 2 above)
Comment (johanstokking) [reply pkg/applicationserver/applicationserver.go:919]: This fix needs to go to `v3.13` right?

**Pull request merged**